spec.html

```diff
@@ -87,80 +48,63 @@ <h1>Uint8Array.prototype.toHex ( )</h1>
 </emu-clause>
 
 <emu-clause id="sec-uint8array.frombase64">
-  <h1>Uint8Array.fromBase64 ( _value_ [ , _options_ ] )</h1>
+  <h1>Uint8Array.fromBase64 ( _string_ [ , _options_ ] )</h1>
+  <p>The <dfn id="standard-base64-alphabet">standard base64 alphabet</dfn> is a List whose elements are the code points corresponding to every letter and number in the Unicode Basic Latin block along with *"+"* and *"/"*; that is, it is StringToCodePoints(*"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/").</p>
   <emu-alg>
-    1. Let _string_ be ? GetStringForBinaryEncoding(_value_).
+    1. If _string_ is not a String, throw a *TypeError* exception.
     1. Let _opts_ be ? GetOptionsObject(_options_).
     1. Let _alphabet_ be ? Get(_opts_, *"alphabet"*).
     1. If _alphabet_ is *undefined*, set _alphabet_ to *"base64"*.
-    1. Set _alphabet_ to ? ToString(_alphabet_).
+    1. If _alphabet_ is not a String, throw a *TypeError* exception.
     1. If _alphabet_ is neither *"base64"* nor *"base64url"*, throw a *TypeError* exception.
-    1. TODO: normalize whitespace / padding here.
-    1. TODO: figure out what the right defaults for whitespace/padding are.
-    1. Let _characters_ be StringToCodePoints(_string_).
-    1. NOTE: StringToCodePoints is used only because RFC 4648 does not produce a sequence of code units. Implementations may be able to base64-decode _string_ directly.
-    1. If _alphabet_ is *"base64"*, then
-      1. If _characters_ cannot result from applying the base64 encoding specified in section 4 of RFC 4648 to some sequence of bytes, throw a *SyntaxError* exception.
-      1. Let _bytes_ be the unique sequence of bytes such that applying the base64 encoding specified in section 4 of RFC 4648 to that sequence would produce _characters_.
+    1. Let _strict_ be ToBoolean(? Get(_opts_, *"strict"*)).
+    1. NOTE: The order of validation and decoding in the algorithm below is not observable. Implementations are encouraged to perform them in whatever order is most efficient, possibly interleaving validation and stripping whitespace with decoding, as long as the behaviour is observably equivalent.
```

Comment:
```suggestion
    1. NOTE: The order of validation and decoding in the algorithm below is not observable except that both must fully complete before invocation of AllocateTypedArray. Implementations are encouraged to perform them in whatever order is most efficient, possibly interleaving validation and stripping whitespace with decoding, as long as the behaviour is observably equivalent.
```

Reply:
Ehhh, I don't think that's necessary. This is a non-normative note.

Reply:
Handling for excessively-long inputs is not generally well-specified at all. Many other algorithms implicitly ask implementations to allocate memory during execution of the algorithm, which is always a fallible operation, and do not specify what happens when that fails, even though in practice it implementations generally throw a RangeError. I don't want to imply there's _more_ restrictions on this algorithm than any other. So I don't want to write down they can't throw a RangeError in the middle of this algorithm; they can, in practice, throw a RangeError in the middle of basically any algorithm.

Reply:
FWIW, it does seem common for observable behavior of implementations to correspond with validity checks preempting allocation failures (see below), and I would like the spec to be clear about that everywhere. But point taken regarding this algorithm not being distinguished in that dimension.
```console
$ eshost -se 'encodeURIComponent("€".repeat(4e8)+"\uDBFF").length'
#### ChakraCore

URIError: The URI to be encoded contains an invalid character

#### JavaScriptCore


#### Moddable XS

Error: memory full

#### SpiderMonkey

URIError: malformed URI sequence

#### V8

URIError: URI malformed
```
(the JavaScriptCore blank seems a silent failure similar to that of e.g. `encodeURIComponent("€".repeat(4e8)).length`—more like XS than like SpiderMonkey and V8)

Reply:
Yeah, definitely agreed that it would be nice to be clear everywhere. I just don't want to do it in this place and not others, when there's nothing really special about this one. Rectifying memory handling is a larger project.

I'm also not totally sure that requiring validity checks up front is what we'd actually want - the case of excessively long inputs is quite rare, and requiring an extra pass prior to decoding just so the right error can get thrown in those cases seems like it would be a shame.

Reply:
> I'm also not totally sure that requiring validity checks up front is what we'd actually want - the case of excessively long inputs is quite rare, and requiring an extra pass prior to decoding just so the right error can get thrown in those cases seems like it would be a shame.

Sure. And that behavior doesn't require an _extra_ pass, but it does require at least _one_ pass (e.g., an implementation could check the remainder of input for validity before deciding to expose allocation failure as an error). Regardless, this discussion at least is settled as non-specific to base64.